Add JsonCondition support

theref opened the pull request:
Implements JsonCondition to match the new condition type added to nucypher - https://github.com/nucypher/nucypher/pull/3656

## What
Adds support for evaluating JSON data directly without HTTP requests. Unlike JsonApiCondition which fetches data from an endpoint, JsonCondition accepts JSON data inline.

## Implementation
- **Schema**: `jsonConditionSchema` with `data` (any type), optional `query` (JSONPath), and `returnValueTest`
- **Class**: `JsonCondition` extends base `Condition` class
- **Factory**: Registered in `ConditionFactory` for deserialization
- **Tests**: 15 unit tests + integration test for backend validation

## Use Cases
- Evaluating static JSON data
- Testing conditions without external dependencies  
- Inline data conditions in multi-condition logic

theref committed:
✨ feat(taco): add JsonCondition schema and class

Add JsonCondition for direct JSON data evaluation without HTTP requests.

Features:
- Accepts JSON data as objects, arrays, primitives, or strings
- Optional JSONPath query support for value extraction
- Context variable resolution in queries and return values
- Zod schema validation with jsonConditionSchema

This parallels the JsonCondition implementation added to nucypher
and provides TypeScript support for the new condition type.

theref committed:
✨ feat(taco): register JsonCondition in condition factory

Add JsonCondition to ConditionFactory.conditionFromProps() to enable
condition creation from serialized JSON condition objects.

This completes the integration of JsonCondition into the taco-web
condition system.

theref committed:
✅ test(taco): add tests for JsonCondition

Add comprehensive test coverage for JsonCondition:

Unit tests (15 tests):
- Schema validation for various data types (object, array, primitives)
- JSONPath query validation (valid and invalid paths)
- Context variable support
- Error handling for invalid condition types

Integration test:
- Backend compatibility validation via Lynx nodes

Includes testJsonConditionObj in test-utils for reusability.

theref commented:
Note for reviewers, the integration test is currently skipped because the corresponding `nucypher` PR hasn't been merged and deployed to lynx

derekpierre commented on packages/taco/src/conditions/schemas/json.ts:
```diff
@@ -0,0 +1,15 @@
+export const jsonConditionSchema = baseConditionSchema.extend({
+  conditionType: z.literal(JsonConditionType).default(JsonConditionType),
+  data: z.any().describe('The JSON data to evaluate. Can be an object, array, primitive value, or JSON string.'),
+  query: jsonPathSchema.optional().describe('Optional JSONPath query to extract a specific value from the data.'),
```
See comments in `nucypher` PR.

derekpierre reviewed (commented):
Just one comment pointing to comments in the `nucypher` PR.

derekpierre commented:
Once the condition structure/values are finalized, you will also need to run the following to include it in the condition-schemas markdown file.

```bash
$ cd packages/taco/

$ pnpm run generate-zod-docs
```

manumonti commented on packages/taco/test/conditions/base/json.test.ts:
```diff
@@ -0,0 +1,104 @@
+          expect(result.error).toBeDefined();
+          const errorMessages = result.error?.errors.map((err) => err.message);
+          expect(
+            errorMessages?.includes('Invalid JSONPath expression'),
```
This test should be changed if we decide to address my suggestion on https://github.com/nucypher/nucypher/pull/3656#discussion_r2393817798

manumonti reviewed (approved):
LGTM! 

Again, good PR, good commit messages, it's a pleasure reviewing this type of PRs.

cygnusv commented on packages/taco/integration-test/condition-lingo.test.ts:
```diff
@@ -173,5 +174,14 @@ describe.skipIf(!process.env.RUNNING_IN_CI)(
     }, 15000);
+
+    test.skip('validate json condition structure', async () => {
+      // TODO: Enable once JsonCondition is deployed to Lynx nodes
```
Is this something to track with an issue?

derekpierre replied:
We can probably unskip this test and accept that it'll fail until `lynx` is updated. In any case, you don't want this PR merged before the `nucypher` PR is merged.

Once the `nucypher`-related PR is merged, we can update `lynx` with the latest from v7.7.x, and then this test can be re-run and will pass, then we can merge this PR.

derekpierre replied:
`lynx` has been updated.

cygnusv commented on packages/taco/test/conditions/base/json.test.ts:
```diff
@@ -0,0 +1,104 @@
+describe('JsonCondition', () => {
+  describe('validation', () => {
+    it.each([
+      [
+        'object',
+        { store: { book: [{ price: 10.5 }] } },
+        '$.store.book[0].price',
+        10.5,
+      ],
+      ['array', [1, 2, 3, 4, 5], '$[2]', 3],
+      ['number', 42, undefined, 42],
+      ['string', 'hello world', undefined, 'hello world'],
+      ['boolean', true, undefined, true],
```
How about adding a JSON-looking string (not a dict) ? e.g. `"{ store: { book: [{ price: 10.5 }] } }"`
Also, it seems the first element of all these test cases (`type`?) is not used.

cygnusv reviewed (approved):
LGTM, just a minor comment on a test case.

theref committed:
✅ test(taco): improve JsonCondition test coverage per code review

- Remove unused first parameter in parameterized test cases
- Add test for JSON-looking strings to verify they remain as strings
- Improve test description formatting to show actual data values

Addresses code review feedback from @cygnusv

theref committed:
✅ test(taco): unskip JsonCondition integration test

theref committed:
♻️ refactor(taco): require JsonCondition data to be context variable

The data field must now be a context variable that resolves to JSON at
decryption time, matching the updated nucypher implementation.

- Update jsonConditionSchema to use contextParamSchema for data field
- Update all tests to use context variable format (:jsonData)
- Add test to verify non-context variable data is rejected
- Update test-utils testJsonConditionObj

theref committed:
✅ test(taco): add JsonCondition encrypt/decrypt integration tests

Add tests that verify JsonCondition works end-to-end with context
variable resolution during encryption and decryption:

- Test successful decrypt when JSON data matches condition
- Test failed decrypt when JSON data doesn't match condition

derekpierre commented:
> Once the condition structure/values are finalized, you will also need to run the following to include it in the condition-schemas markdown file.
> 
> ```shell
> $ cd packages/taco/
> 
> $ pnpm run generate-zod-docs
> ```

^ @theref , reminder to do the above before merging.

theref committed:
📝 docs(taco): add JsonCondition to schema documentation

Export jsonConditionSchema in zod-doc-gen and regenerate condition
schemas markdown to include JsonCondition documentation.

derekpierre reviewed (approved):
🎸
